tower-http/src/request_id.rs

```diff
@@ -0,0 +1,575 @@
+//! #             .fetch_add(1, Ordering::SeqCst)
+//! #             .to_string()
+//! #             .parse()
+//! #             .unwrap();
+//! #         Some(RequestId::new(request_id))
+//! #     }
+//! # }
+//!
+//! let svc = ServiceBuilder::new()
+//!     // make sure to set request ids before the request reaches `TraceLayer`
+//!     .set_x_request_id(MyMakeRequestId::default())
+//!     // log requests and responses
+//!     .layer(
+//!         TraceLayer::new_for_http()
+//!             .make_span_with(DefaultMakeSpan::new().include_headers(true))
+//!             .on_response(DefaultOnResponse::new().include_headers(true))
+//!     )
+//!     // propagate the header to the response before the response reaches `TraceLayer`
+//!     .propagate_x_request_id()
+//!     .service(handler);
+//! #
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! # Using `UUID`s
+//!
+//! Generating request ids from [`Uuid`]s can be done like so:
+//!
+//! ```
+//! use tower_http::ServiceBuilderExt;
+//! use http::{Request, Response, header::HeaderName};
+//! use tower::{Service, ServiceExt, ServiceBuilder};
+//! use tower_http::request_id::{
+//!     SetRequestIdLayer, PropagateRequestIdLayer, MakeRequestId, RequestId,
+//! };
+//! use uuid::Uuid;
+//! # use hyper::Body;
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! # let handler = tower::service_fn(|request: Request<Body>| async move {
+//! #     Ok::<_, std::convert::Infallible>(Response::new(request.into_body()))
+//! # });
+//!
+//! #[derive(Clone, Copy)]
+//! struct MakeRequestUuid;
+//!
+//! impl MakeRequestId for MakeRequestUuid {
+//!     fn make_request_id<B>(&mut self, request: &Request<B>) -> Option<RequestId> {
+//!         let request_id = Uuid::new_v4()
+//!             .to_string()
+//!             .parse()
+//!             .unwrap();
+//!         Some(RequestId::new(request_id))
+//!     }
+//! }
+//!
+//! let mut svc = ServiceBuilder::new()
+//!     .set_x_request_id(MakeRequestUuid)
+//!     .propagate_x_request_id()
+//!     .service(handler);
+//!
+//! let request = Request::new(Body::empty());
+//! let response = svc.ready().await?.call(request).await?;
+//!
+//! assert!(response.headers().get("x-request-id").is_some());
+//! #
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! [`ServiceBuilderExt`]: crate::ServiceBuilderExt
+//! [`Uuid`]: https://crates.io/crates/uuid
+//! [`Trace`]: crate::trace::Trace
+
+// NOTE: when uuid 1.0 is shipped we can include a `MakeRequestId` that uses that
+// See https://github.com/uuid-rs/uuid/issues/113
+
+use http::{
+    header::{HeaderName, HeaderValue},
+    Request, Response,
+};
+use pin_project_lite::pin_project;
+use std::task::{Context, Poll};
+use std::{future::Future, pin::Pin};
+use tower_layer::Layer;
+use tower_service::Service;
+
+pub(crate) const X_REQUEST_ID: &str = "x-request-id";
+
+/// Trait for producing [`RequestId`]s.
+///
+/// Used by [`SetRequestId`].
+pub trait MakeRequestId {
```

Comment:
Might as well `MakeRequestId: Clone` since it is anyway required for it to be useful

Reply:
It's only required for the layer and not the service. So not strictly required.